cpp/src/reductions/histogram_helpers.hpp

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column_view.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/table/table_view.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <optional>
+
+namespace cudf::reduction::detail {
+
+/**
+ * @brief Compute the histogram for the input table.
+ *
+ * This is equivalent to do a distinct count for each unique row in the input.
+ *
+ * @param input The input table to compute histogram
+ * @param partial_counts An optional column containing count for each row
+ * @param output_dtype The type to store count value
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate memory of the returned objects
+ * @return A pair of array contains the (stable-order) indices of the distinct rows in the input
+ * table, and their corresponding distinct counts
+ */
+std::pair<rmm::device_uvector<size_type>, std::unique_ptr<column>> table_histogram(
```

Comment:
Not a hard requirement, but I feel this should be just another overload of `histogram` but taking `table_view` instead of `column_view`. `table_histogram` does not follow a consistent naming like in other places we the table overload of `contains` as `contains_table` instead of `table_contains`.

Reply:
Then how about `histogram_table`? This function is intended to be used across reduction and groupby so I want to differentiate its name from `histogram` in reduction.

Reply:
That's fine

Reply:
Change it to `compute_row_frequencies` since it indeed doesn't output a complete histogram but just the intermediate data to build a histogram.